rule.php

```diff
@@ -0,0 +1,157 @@
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Implementaton of the quizaccess_seb plugin.
+ *
+ * @package    quizaccess_seb
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/mod/quiz/accessrule/accessrulebase.php');
+
+class quizaccess_seb extends quiz_access_rule_base {
+
+    /**
+     * quizaccess_seb constructor.
+     */
+    public function __construct($quizobj, $timenow) {
+        parent::__construct($quizobj, $timenow);
+    }
+
+    /**
+     * @param quiz $quizobj
+     * @param int $timenow
+     * @param bool $canignoretimelimits
+     * @return quiz_access_rule_base|quizaccess_seb|null
+     */
+    public static function make(quiz $quizobj, $timenow, $canignoretimelimits) {
+        return new self($quizobj, $timenow);
+    }
+
+    /**
+     * Add extra setting elements to the quiz settings.
+     *
+     * @param mod_quiz_mod_form $quizform
+     * @param MoodleQuickForm $mform
+     */
+    public static function add_settings_form_fields(mod_quiz_mod_form $quizform, MoodleQuickForm $mform) {
+    }
+
+    /**
+     * Validate the form elements when form is submitted.
+     *
+     * @param array $errors
+     * @param array $data
+     * @param array $files
+     * @param mod_quiz_mod_form $quizform
+     * @return array
+     */
+    public static function validate_settings_form_fields(array $errors,
+                                                         array $data, $files, mod_quiz_mod_form $quizform) {
+        return $errors;
+    }
+
+    /**
+     * Hook to handle saving the quiz settings for plugin.
+     *
+     * @param object $quiz
+     */
+    public static function save_settings($quiz) {
+    }
+
+    /**
+     * Delete quiz settings from DB and related files from storage.
+     *
+     * @param object $quiz
+     */
+    public static function delete_settings($quiz) {
+    }
+
+    /**
+     * Return the SQL needed to load all the settings from all the access plugins in one DB query.
+     *
+     * Array to return contains three elements - [fields, joins, params].
+     *
+     * @param int $quizid
+     * @return array
+     */
+    public static function get_settings_sql($quizid) {
```

Comment:
I think the general rule of thumb (not s standard yet) to cast return type for the functions

Reply:
Ah thanks, assuming you mean by casting  `function() : array {}`, I've added the return time to function sig where possible. All of these methods are stubbed from the base class.